### DISCUSSION TITLE
Threading in seleniumbase

### DISCUSSION THREAD

**Discussion by misterpeople** [seleniumbase/SeleniumBase#2835] [Q&A] [answered]
Hey, so I'm trying to bypass Cloudflare's "Verify you are human" CAPTCHA, and it works sometimes. However, when it doesn't work and asks us to click on the verify box, I encounter an issue. Once it clicks on the box, it never bypasses Cloudflare, and I'm stuck on the page without being able to proceed. Am I doing something wrong here?

'''
import threading

# Multi-threading for handling multiple accounts
def main():
    conn = get_db_connection()
    cursor = conn.cursor()
    cursor.execute('SELECT * FROM accounts')
    accounts = cursor.fetchall()
    conn.close()

    threads = []
    for account in accounts:
        t = threading.Thread(target=handle_login, args=(account,))
        threads.append(t)
        t.start()
        time.sleep(int(os.getenv('LOGIN_DELAY', 5)))  # Control delay between logins

    for t in threads:
        t.join()
'''
I wrote a simple threading code for it, will Selenmiumbase support it or am I totally wrong? Im new coder so sorry if its an annoying question .

**Comment by mdmintz** [comment, marked as answer]
Duplicate of https://github.com/seleniumbase/SeleniumBase/issues/2828#issuecomment-2147511383

See the example that uses `ThreadPoolExecutor` with UC Mode:

```python
import sys
from concurrent.futures import ThreadPoolExecutor
from seleniumbase import Driver
sys.argv.append("-n")  # Tell SeleniumBase to do thread-locking as needed

def launch_driver(url):
    driver = Driver(uc=True)
    try:
        driver.get(url=url)
        driver.sleep(2)
    finally:
        driver.quit()

urls = ['https://seleniumbase.io/demo_page' for i in range(3)]
with ThreadPoolExecutor(max_workers=len(urls)) as executor:
    for url in urls:
        executor.submit(launch_driver, url)
```

Or use `pytest` multithreading with `pytest-xdist` and the `parameterized` library: https://github.com/seleniumbase/SeleniumBase/discussions/2394#discussioncomment-7969332

**Reply by misterpeople** [reply]
thanks i will check, It wont work the way I did it with threading?
May i understand the differences?

**Reply by mdmintz** [reply]
Only `ThreadPoolExecutor` as described and `pytest` multithreading as described are officially supported.
If there are any other ways that work, you'll have to find out on your own.

**Reply by misterpeople** [reply]
I see. I need to be able to control a single thread as well as multiple threads, For example, the command /off will shut them all down, and each thread will have its own settings (same tasks). Clients may want one out of three choices, while another client may want all three choices, also at anytime there could be 1 thread or even 20 threads if I add more accounts. Is it still possible with ThreadPoolExecutor or pytest?

**Reply by mdmintz** [reply]
Sounds like you'll have to experiment on your own. I don't maintain either of those, but I do use them.
